### ISSUE TITLE
Portal config error: 512 <unknown status code>

### ISSUE THREAD

**Issue by joshuakirsch** [yuezk/GlobalProtect-openconnect#327]
Hi,

I keep recieving this error when trying to access the VPN. Others have been able to access the VPN with the same method on different computers and networks. (I've removed identifying info from the CLI output below)

```
(base) bcem-09@HP-ProDesk:~$ gpclient connect 
[2024-02-21T14:40:24Z INFO  gpclient::cli] gpclient started: 2.0.0 (2024-02-05)
[2024-02-21T14:40:24Z INFO  gpapi::portal::prelogin] Prelogin with user_agent: PAN GlobalProtect
Enter login credentials (Portal: )
> Username: 
> Password: 
[2024-02-21T14:40:35Z INFO  gpapi::portal::config] Portal config, user_agent: PAN GlobalProtect
[2024-02-21T14:41:00Z INFO  gpclient::connect] Failed to connect portal with prelogin: Portal config error: 512 <unknown status code>

Error: Portal config error: 512 <unknown status code>
```

**Comment by yuezk** [comment]
Error code 512 is a general code, your credentials may not be correct. Have you tried the GUI version?

**Comment by joshuakirsch** [comment]
Yes and I got the same error

**Comment by yuezk** [comment]
Are you using the email as the username? If so, please try to use the email handle as the username.

**Comment by yuezk** [comment]
@joshuakirsch `openconnect` supports username/password authentication, you could try it out to see if it works for you.

**Comment by joshuakirsch** [comment]
Hi, I changed my default browser to Brave and added a 2-factor authentication step. I am now running into a new error:

(base) bcem-09@HP-ProDesk:~$ gpclient connect 
[2024-02-23T15:21:03Z INFO  gpclient::cli] gpclient started: 2.0.0 (2024-02-05)
[2024-02-23T15:21:03Z INFO  gpapi::portal::prelogin] Prelogin with user_agent: PAN GlobalProtect
Enter login credentials (Portal: )
> Username: 
> Password: 
[2024-02-23T15:21:20Z INFO  gpapi::portal::config] Portal config, user_agent: PAN GlobalProtect
[2024-02-23T15:21:36Z INFO  gpapi::gateway::login] Gateway login, user_agent: PAN GlobalProtect
[2024-02-23T15:21:36Z INFO  gpclient::connect] Gateway login failed: Gateway login error: 512 <unknown status code>
[2024-02-23T15:21:36Z INFO  gpapi::portal::prelogin] Prelogin with user_agent: PAN GlobalProtect
Enter login credentials (Gateway: )
> Username: 
> Password: 
[2024-02-23T15:21:56Z INFO  gpapi::gateway::login] Gateway login, user_agent: PAN GlobalProtect
[2024-02-23T15:22:02Z INFO  openconnect::ffi] openconnect version: v8.20-1
[2024-02-23T15:22:02Z INFO  openconnect::ffi] User agent: PAN GlobalProtect
[2024-02-23T15:22:02Z INFO  openconnect::ffi] VPNC script: /usr/share/vpnc-scripts/vpnc-script
[2024-02-23T15:22:02Z INFO  openconnect::ffi] OS: linux
[2024-02-23T15:22:02Z INFO  openconnect::ffi] CSD_USER: 1000
[2024-02-23T15:22:02Z INFO  openconnect::ffi] CSD_WRAPPER: (null)
[2024-02-23T15:22:02Z INFO  openconnect::ffi] POST https://amc-vpn-linux.ucdenver.edu/ssl-vpn/getconfig.esp
[2024-02-23T15:22:02Z INFO  openconnect::ffi] Connected to 140.226.8.251:443
[2024-02-23T15:22:02Z INFO  openconnect::ffi] SSL negotiation with 
[2024-02-23T15:22:03Z INFO  openconnect::ffi] Connected to HTTPS on  with ciphersuite (TLS1.2)-(ECDHE-SECP256R1)-(RSA-SHA256)-(AES-256-GCM)
[2024-02-23T15:22:03Z INFO  openconnect::ffi] Tunnel timeout (rekey interval) is 180 minutes.
[2024-02-23T15:22:03Z INFO  openconnect::ffi] Idle timeout is 180 minutes.
[2024-02-23T15:22:03Z WARN  openconnect::ffi] GlobalProtect IPv6 support is experimental. Please report results to <openconnect-devel@lists.infradead.org>.
[2024-02-23T15:22:03Z WARN  openconnect::ffi] No MTU received. Calculated 1422 for ESP tunnel
[2024-02-23T15:22:03Z INFO  openconnect::ffi] POST 
[2024-02-23T15:22:03Z WARN  openconnect::ffi] WARNING: Server asked us to submit HIP report with md5sum c43dededcba01cb45d92a1110105d82e.
        VPN connectivity may be disabled or limited without HIP report submission.
        You need to provide a --csd-wrapper argument with the HIP report submission script.
[2024-02-23T15:22:03Z INFO  openconnect::ffi] ESP session established with server
[2024-02-23T15:22:03Z INFO  openconnect::ffi] ESP tunnel connected; exiting HTTPS mainloop.
mkdir: cannot create directory ‘/var/run/vpnc’: Permission denied
[2024-02-23T15:22:04Z WARN  openconnect::ffi] Failed to bind local tun device (TUNSETIFF): Operation not permitted
[2024-02-23T15:22:04Z WARN  openconnect::ffi] To configure local networking, openconnect must be running as root
    See https://www.infradead.org/openconnect/nonroot.html for more information
mkdir: cannot create directory ‘/var/run/vpnc’: Permission denied
[2024-02-23T15:22:04Z WARN  openconnect::ffi] Failed to bind local tun device (TUNSETIFF): Operation not permitted
[2024-02-23T15:22:04Z WARN  openconnect::ffi] To configure local networking, openconnect must be running as root
    See https://www.infradead.org/openconnect/nonroot.html for more information
Set up tun device failed
[2024-02-23T15:22:04Z INFO  openconnect::ffi] POST https://amc-vpn-linux.ucdenver.edu/ssl-vpn/logout.esp
[2024-02-23T15:22:04Z INFO  openconnect::ffi] SSL negotiation with 
[2024-02-23T15:22:05Z INFO  openconnect::ffi] Connected to HTTPS on  with ciphersuite (TLS1.2)-(ECDHE-SECP256R1)-(RSA-SHA256)-(AES-256-GCM)
[2024-02-23T15:22:05Z INFO  openconnect::ffi] Logout successful.
[2024-02-23T15:22:05Z INFO  openconnect::ffi] openconnect_mainloop returned -5, exiting

**Comment by yuezk** [comment]
@joshuakirsch Please use `sudo` to run `gpclient`.

**Comment by alfredopalhares** [comment]
Hello, 

Thank you for working on such tool.  

I am having the same error, is there a way to force the tool to print out the response ?

**Comment by yuezk** [comment]
As I remember, 512 status code is a generic error code for GP endpoint, and the response body is empty. Anyway, I will print the response when this error occurred.

**Comment by yuezk** [comment]
Released in 2.1.2, which will output the details of the error response.
